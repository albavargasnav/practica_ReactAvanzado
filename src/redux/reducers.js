import { AUTH_LOGIN, AUTH_LOGOUT, ADVERTS_LOADED } from './types';

export const defaultState = {
  auth: false,
  adverts: [],
};

// export default function reducer(state = defaultState, action) {
//   switch (action.type) {
//     case AUTH_LOGIN:
//       return { ...state, auth: true };
//     case AUTH_LOGOUT:
//       return { ...state, auth: false };
//     case TWEETS_LOADED:
//       return { ...state, tweets: action.payload };
//     default:
//       return state;
//   }
// }

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  if (action.type === ADVERTS_LOADED) {
    return action.payload;
  }
  return state;
}

// export default function combinedReducer(state = defaultState, action) {
//   return {
//     auth: auth(state.auth, action),
//     tweets: tweets(state.tweets, action),
//   };
// }