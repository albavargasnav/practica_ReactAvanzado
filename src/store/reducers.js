import {
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_CREATED_SUCCESS,
  ADVERT_LOADED_SUCCESS,
  TAGS_REQUEST,
  TAGS_SUCCESS,
  TAGS_FAILURE,
  UI_RESET_ERROR,
} from './types';

export const defaultState = {
  auth: false,
  adverts: {
    areLoaded: false,
    data: [],
  },
  ui: {
    isLoading: false,
    error: null,
    tagsLoading: false,
    tagsError: null,
  },
};

// export default function reducer(state = defaultState, action) {
//   switch (action.type) {
//     case AUTH_LOGIN:
//       return { ...state, auth: true };
//     case AUTH_LOGOUT:
//       return { ...state, auth: false };
//     case ADVERTS_LOADED:
//       return { ...state, adverts: action.payload };
//     default:
//       return state;
//   }
// }

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return {
        ...state,
        loading: false,
        isAuthenticated: true,
        token: action.payload.token, 
      };
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  if (action.type === ADVERTS_LOADED_SUCCESS) {
    return { areLoaded: true, data: action.payload };
  }
  if (action.type === ADVERT_LOADED_SUCCESS) {
    return { ...state, data: [action.payload] };
  }
  if (action.type === ADVERT_CREATED_SUCCESS) {
    // return { ...state, data: [action.payload, ...state.data] };
    return { ...state, data: [action.payload].concat(state.data) };
  }
  return state;
}

export function ui(state = defaultState.ui, action) {
  if (action.type === TAGS_REQUEST) {
    return {
      ...state,
      tagsLoading: true,
      tagsError: null,
    };
  }
  if (action.type === TAGS_SUCCESS) {
    return {
      ...state,
      tagsLoading: false,
    };
  }
  if (action.type === TAGS_FAILURE) {
    return {
      ...state,
      tagsLoading: false,
      tagsError: action.payload,
    };
  }
  if (action.error) {
    return { isLoading: false, error: action.payload };
  }

  if (/_REQUEST$/.test(action.type)) {
    return { isLoading: true, error: null };
  }

  if (/_SUCCESS$/.test(action.type)) {
    return { isLoading: false, error: null };
  }

  if (action.type === UI_RESET_ERROR) {
    return { ...state, error: null };
  }

  return state;
}


export default function combinedReducer(state = defaultState, action) {
  return {
    auth: auth(state.auth, action),
    adverts: adverts(state.adverts, action),
    ui: ui(state.ui, action),
  };
}