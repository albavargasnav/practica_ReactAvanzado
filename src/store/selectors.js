export const getIsLogged = state => state.auth;

export const getAdverts = state => state.adverts;

// export const getAdvert = (state, advertId) =>
//   getAdverts(state).find(advert => advert.id === +advertId);

export const getAdvert = advertId => state =>
  getAdverts(state).find(advert => advert.id === Number(advertId));