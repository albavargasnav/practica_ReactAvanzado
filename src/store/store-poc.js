const INCREMENT = 'INCREMENT';
const DECREMENT = 'DECREMENT';

function reducer(state = 0, action) {
  switch (action.type) {
    case INCREMENT:
      return state + action.payload;
    case DECREMENT:
      return state - action.payload;
    default:
      return state;
  }
}

function createStore(reducer, preloadedState) {
  let state = preloadedState;
  let listeners = [];
  const getState = () => state;
  const dispatch = action => {
    state = reducer(state, action);
    listeners.forEach(listener => listener());
  };
  const subscribe = listener => {
    listeners.push(listener);
    return function () {
      listeners = listeners.filter(l => l !== listener);
    };
  };
  dispatch({ type: 'INIT' });
  return {
    getState,
    dispatch,
    subscribe,
  };
}

const increment = step => ({
  type: INCREMENT,
  payload: step ?? 1,
});

const decrement = step => ({
  type: DECREMENT,
  payload: step ?? 1,
});

const store = createStore(reducer);
const showState = () => console.log(store.getState());
showState();

const unsubscribe = store.subscribe(showState);
// store.dispatch({ type: INCREMENT, payload: 2 });
store.dispatch(increment(2));
//store.dispatch({ type: DECREMENT, payload: 3 });
store.dispatch(decrement(3));

store.dispatch({ type: DECREMENT });
store.dispatch({ type: DECREMENT });
store.dispatch({ type: DECREMENT });
store.dispatch({ type: INCREMENT });
unsubscribe();