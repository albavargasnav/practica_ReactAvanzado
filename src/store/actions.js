import { areAdvertsLoaded, getAdvert } from './selectors';
import {
    AUTH_LOGIN_FAILURE,
    AUTH_LOGIN_REQUEST,
    AUTH_LOGIN_SUCCESS,
    AUTH_LOGOUT,
    ADVERTS_LOADED_FAILURE,
    ADVERTS_LOADED_REQUEST,
    ADVERTS_LOADED_SUCCESS,
    ADVERT_LOADED_FAILURE,
    ADVERT_LOADED_SUCCESS,
    ADVERT_LOADED_REQUEST,
    UI_RESET_ERROR,
    ADVERT_CREATED_SUCCESS,
    ADVERT_CREATED_REQUEST,
    ADVERT_CREATED_FAILURE,
  } from './types';

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});

export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  error: true,
  payload: error,
});

export const authLogin =
  credentials =>
  async (dispatch, _getState, { auth }) => {
    dispatch(authLoginRequest());
    try {
      await auth.login(credentials);
      // Logged in
      dispatch(authLoginSuccess());
    } catch (error) {
      dispatch(authLoginFailure(error));
      throw error;
    }
  };

export const authLogout = () => ({
    type: AUTH_LOGOUT,
});

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = adverts => ({
  type: ADVERTS_LOADED_SUCCESS,
    payload: adverts,
});

export const advertsLoadedFailure = error => ({
  type: ADVERTS_LOADED_FAILURE,
  error: true,
  payload: error,
});

export const advertsLoaded =
  () =>
  async (dispatch, getState, { adverts: advertsService }) => {
    if (areAdvertsLoaded(getState())) {
      return;
    }
    dispatch(advertsLoadedRequest());
    try {
      const adverts = await advertsService.getLatestAdverts();
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
    }
  };

  export const advertLoadedRequest = () => ({
    type: ADVERT_LOADED_REQUEST,
  });
  
  export const advertLoadedSuccess = advert => ({
    type: ADVERT_LOADED_SUCCESS,
    payload: advert,
  });
  
  export const advertLoadedFailure = error => ({
    type: ADVERT_LOADED_FAILURE,
    error: true,
    payload: error,
  });
  
  export const advertLoad =
    advertId =>
    async (dispatch, getState, { adverts: advertsService }) => {
      const isLoaded = getAdvert(advertId)(getState());
      if (isLoaded) {
        return;
      }
      dispatch(advertLoadedRequest());
      try {
        const advert = await advertsService.getAdvert(advertId);
        dispatch(advertLoadedSuccess(advert));
      } catch (error) {
        dispatch(advertLoadedFailure(error));
        throw error;
      }
    };

    export const advertCreateRequest = () => ({
      type: ADVERT_CREATED_REQUEST,
    });
    
    export const advertCreateSuccess = advert => ({
      type: ADVERT_CREATED_SUCCESS,
      payload: advert,
    });
    
    export const advertCreateFailure = error => ({
      type: ADVERT_CREATED_FAILURE,
      error: true,
      payload: error,
    });
    
    export const advertCreate =
      advert =>
      async (dispatch, _getState, { adverts: advertsService }) => {
        dispatch(advertCreateRequest());
        try {
          const { id } = await advertsService.createAdvert(advert);
          const createdAdvert = await advertsService.getAdvert(id);
          dispatch(advertCreateSuccess(createdAdvert));
          return createdAdvert;
        } catch (error) {
          dispatch(advertCreateFailure(error));
          throw error;
        }
      };

export const uiResetError = () => ({
    type: UI_RESET_ERROR,
  });

