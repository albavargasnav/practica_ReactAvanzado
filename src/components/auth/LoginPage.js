import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import Button from '../shared/Button';
import FormField from '../shared/FormField';
import { login } from './service';
import './LoginPage.css';
import { useLocation, useNavigate } from 'react-router-dom';
import { useRef } from 'react';
import { useEffect } from 'react';
import {
  authLoginFailure,
  authLoginRequest,
  authLoginSuccess,
  uiResetError,
} from '../../store/actions';
import { getUi } from '../../store/selectors';


function LoginPage() {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const location = useLocation();
  const { isLoading, error } = useSelector(getUi);
  const renders = useRef(0);

  useEffect(() => {
    renders.current++;
    console.log(renders.current, ' times rendered');
  });

  const [credentials, setCredentials] = useState({
        email: '',
        password: '',
  });

  const [rememberMe, setRememberMe] = useState(false);

  const resetError = () => {
    dispatch(uiResetError());
  };

  const onLogin = () => dispatch(authLoginSuccess());


  const handleSubmit = async event => {
    event.preventDefault();
    dispatch(authLoginRequest());
    try {
      await login(credentials, rememberMe);
      onLogin();
      const to = location.state?.from?.pathname || '/';
      navigate(to);
    } catch (error) {
      dispatch(authLoginFailure(error));
    }
  };

  const handleChange = event => {

    setCredentials({
      ...credentials,
      [event.target.name]: event.target.value,
    });
  };

  const handleRememberMeChange = event => {
    setRememberMe(event.target.checked);
  };

  const buttonDisabled =
    isLoading || !credentials.email || !credentials.password;

  return (
    <div className="loginPage">
      <h1 className="loginPage-title">Log in to Adverts</h1>
      <form onSubmit={handleSubmit}>
        <FormField
          type="text"
          name="email"
          label="Email"
          className="loginForm-field"
          onChange={handleChange}
          value={credentials.username}
          autofocus
        />
        <FormField
          type="password"
          name="password"
          label="password"
          className="loginForm-field"
          onChange={handleChange}
          value={credentials.password}
        />

        <div className="rememberMe">
          <input
            type="checkbox"
            id="rememberMe"
            name="rememberMe"
            checked={rememberMe}
            onChange={handleRememberMeChange}
          />
          <label htmlFor="rememberMe">Remember me</label>
        </div>
        <Button
          type="submit"
          variant="primary"
          className="loginForm-submit"
          disabled={buttonDisabled}
        >
          Log in
        </Button>
      </form>
      {error && (
        <div onClick={resetError} className="loginPage-error">
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
