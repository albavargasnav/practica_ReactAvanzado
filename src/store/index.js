import { createStore, combineReducers, applyMiddleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';

import thunk from 'redux-thunk';

import * as service from '../service';

import * as reducers from './reducers';
import * as actionCreators from './actions'

import { authLoginSuccess } from './actions';


const reducer = combineReducers(reducers);

const composeEnhancers = composeWithDevTools ({
    actionCreators,
})
  
  const logger = store => next => action => {
    if (action.type) {
      console.group(action.type);
      console.info('dispatching', action, store.getState());
      const result = next(action);
      console.log('next state', store.getState());
      console.groupEnd();
      return result;
    }
    return next(action);
  };
  
  const timestamp = () => next => action => {
    return next({
      ...action,
      meta: {
        ...action.meta,
        timestamp: Date.now(),
      },
    });
  };
  
  const failureRedirects = (router, redirectsMap) => () => next => action => {
    const result = next(action);
  
    if (action.error) {
      const redirect = redirectsMap[action.payload.status];
      if (redirect) {
        router.navigate(redirect);
      }
    return result;
  };
}
  
  export default function configureStore(preloadedState, { router }) {
    const middleware = [
      thunk.withExtraArgument({ service, router }),
      timestamp,
      logger,
    ];
  
    const store = createStore(
        reducer,
        preloadedState,
        composeEnhancers(applyMiddleware(...middleware)),
      );
    
      const token = localStorage.getItem('token');
      if (token) {
        store.dispatch(authLoginSuccess());
      }

    return store;
  }


