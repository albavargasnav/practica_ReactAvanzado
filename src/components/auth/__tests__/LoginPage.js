import { fireEvent, render, screen } from '@testing-library/react';
import LoginPage from '../LoginPage';
import { Provider } from 'react-redux';
import { defaultState } from '../../../store/reducers';
import { authLogin, uiResetError } from '../../../store/actions';

jest.mock('../../../store/actions');

describe('LoginPage', () => {
  const renderComponent = (error = null) => {
    const store = {
      getState: () => {
        const state = defaultState;
        state.ui.error = error;

        return state;
      },
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );
  };

  test('snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch authLogin action', () => {
    const email = 'david';
    const password = '1234';

    renderComponent();

    const emailInput = screen.getByLabelText(/Email/);
    const passwordInput = screen.getByLabelText(/password/);
    const submitButton = screen.getByRole('button');

    expect(submitButton).toBeDisabled();

    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });

    expect(submitButton).toBeEnabled();

    fireEvent.click(submitButton);

    expect(authLogin).toHaveBeenCalledWith({ email, password});
    console.log (email, password)
  });

  test('should display an error', () => {
    const error = { message: 'Unauthorized' };
    renderComponent(error);

    const errorElement = screen.getByText(error.message);
    expect(errorElement).toBeInTheDocument();

    fireEvent.click(errorElement);

    expect(uiResetError).toHaveBeenCalled();
  });
});