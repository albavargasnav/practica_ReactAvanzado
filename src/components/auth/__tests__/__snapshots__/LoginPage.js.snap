// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in to Adverts
    </h1>
    <form>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            Email
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            password
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <div
        class="rememberMe"
      >
        <input
          id="rememberMe"
          name="rememberMe"
          type="checkbox"
        />
        <label
          for="rememberMe"
        >
          Remember me
        </label>
      </div>
      <button
        class="sc-beqWaB fHPURa loginForm-submit"
        disabled=""
        type="submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;
